toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,329 @@
+        "oauth_client": "ssh-proxy"
+      }
+	}
+  }
+}
+```
+- The "/" endpoint MUST NOT return the legacy SSH info, if the legacy SSH feature is disabled.
+
+- The endpoint MUST return an object for ssh-over-websocket if the feature is enabled:
+
+```
+{
+  "links": {
+	(...)
+    "app_ssh_ws": {
+      "href": "wss://ssh.cf.eu10.hana.ondemand.com",
+      "meta": {
+        "host_key_fingerprint": "TD1dRQINLi2KxilVLzAI8tXB2h8MP79oyVJnUwshjdc",
+        "oauth_client": "ssh-proxy"
+      }
+	}
+  }
+}
+```
+
+- If both legacy SSH and SSH-over-WebSocket features are enabled, the "/" endpoint MUST return info objects for both features:
+
+```
+{
+  "links": {
+	(...)
+    "app_ssh": {
+      "href": "ssh.cf.eu10.hana.ondemand.com:2222",
+      "meta": {
+        "host_key_fingerprint": "TD1dRQINLi2KxilVLzAI8tXB2h8MP79oyVJnUwshjdc",
+        "oauth_client": "ssh-proxy"
+      }
+	}
+
+    "app_ssh_ws": {
+      "href": "wss://ssh.cf.eu10.hana.ondemand.com",
+      "meta": {
+        "host_key_fingerprint": "TD1dRQINLi2KxilVLzAI8tXB2h8MP79oyVJnUwshjdc",
+        "oauth_client": "ssh-proxy"
+      }
+	}
+  }
+}
+```
+
+- CF CLI MUST prefer the `app_ssh_ws` link over the `app_ssh` link if present.
+- CF CLI MAY fall back to `app_ssh` link if `app_ssh_ws` did not connect successfully and both are present.
+- When using `app_ssh_ws`, the CF CLI MUST wrap the SSH connection in a WebSocket connection:
+
+```
+func (c *secureShell) Connect(opts *options.SSHOptions) error {
+	err := c.validateTarget(opts)
+	if err != nil {
+		return err
+	}
+
+	clientConfig := &ssh.ClientConfig{
+		User: fmt.Sprintf("cf:%s/%d", c.app.GUID, opts.Index),
+		Auth: []ssh.AuthMethod{
+			ssh.Password(c.token),
+		},
+		HostKeyCallback: fingerprintCallback(opts, c.sshEndpointFingerprint),
+	}
+
+	// Wrap SSH in WebSocket if possible
+	var secureClient SecureClient
+	if c.sshWsEndpoint != "" {
+		secureClient, err = tunnelSSHThruWebSocket(c.sshWsEndpoint, clientConfig)
+	}
+
+	if err != nil {
+		// Fall back to SSH over TCP
+		secureClient, err = c.secureDialer.Dial("tcp", c.sshEndpoint, clientConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	c.secureClient = secureClient
+	c.opts = opts
+	return nil
+}
+```
+Code adapted from [cloudfoundry/cli ssh.go](https://github.com/cloudfoundry/cli/blob/main/cf/ssh/ssh.go#L119)
+
+The `sshWsEndpoint` would have to be fetched from the `app_ssh_ws.href` field of the "/" endpoint of CAPI.
+
+
+A `tunnelSSHThruWebSocket` function could look something like this:
+```
+func tunnelSSHThruWebSocket(url string, config *ssh.ClientConfig) (SecureClient, error) {
+	wsConn, err := websocket.Dial(url, "", url)
+	if err != nil {
+		return nil, err
+	}
+
+	c, chans, reqs, err := ssh.NewClientConn(wsConn, url, config)
+	if err != nil {
+		return nil, err
+	}
+	sshClient := ssh.NewClient(c, chans, reqs)
+	return &secureClient{client: sshClient}, nil
+}
+```
+
+- The rest of CF CLI SHOULD be unchanged and work as before.
```

Comment:
Question: Local port forwarding (e.g. to a DB service or to JMX port of my Java app) would be supported as well, correct?
Example: `cf ssh -N -L 5000:localhost:5000 myapp`

Reply:
This is a good additional example of why interoperability with OpenSSH is crucial (see the [parallel discussion](https://github.com/cloudfoundry/community/pull/807#discussion_r1559013471)).

I think such cases should be supported via a tunnel command, e.g.:

```
ssh -o ProxyCommand='cf tunnel-ssh my-app' ... ???@???
```

Otherwise we'd have to re-implement major parts of the OpenSSH client which I don't think is reasonable.

Users could configure their ssh client to automatically use the proxy command when connecting to an app to avoid having to type out the options each time.

Reply:
Yes, the features of SSH, including port forwarding should be unaffected as the WebSocket is handled transparently in the `ProxyCommand` option. Similar to what you would get with any other proxy like SOCKS5 etc.
The SSH protocol does not "know" it's being run over a proxy.